IPython/core/magics/execution.py

```diff
@@ -1074,7 +1075,16 @@ def timeit(self, line='', cell=None):
         tc = clock()-t0
 
         ns = {}
-        exec(code, self.shell.user_ns, ns)
+        glob = self.shell.user_ns
+        # handles global vars with same name as local vars. We store them in conflict_globs.
+        if local_ns is not None:
+            conflict_globs = {}
+            for var_name, var_val in glob.items():
+                if var_name in local_ns:
+                    conflict_globs[var_name] = var_val
+            glob.update(local_ns)
+            
+        exec(code, glob, ns)
```

Comment:
I tried using `ns.merge(local_ns)`, but again it leads to same problem. The vars in `ns` are not in `locals()` inside `inner()`, but they are in local namespace, if we access them outside `inner`. Hence, same problem persists. 
If we somehow get rid of that extra stack_frame due to `inner` then we can retain our standard approach.